File: projects/von.ts/von/src/format.ts
import { VonError } from "./error.js";
import { isIdentifierContinue, isIdentifierStart } from "./lexical.js";
import { isVonObject, type VonValue } from "./value.js";

/** Serialize a value to compact, deterministic VON text. */
export function stringify(value: VonValue): string {
  return formatValue(assertVonValue(value), undefined, 0);
}

/** Serialize a value to deterministic, two-space-indented VON text. */
export function stringifyIndented(value: VonValue, indentWidth = 2): string {
  return formatValue(assertVonValue(value), indentWidth, 0);
}

function assertVonValue(value: VonValue): VonValue {
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new VonError("VON 不支持浮点数或超出安全整数范围的数值");
    }
  }
  return value;
}

function formatValue(value: VonValue, indent: number | undefined, depth: number): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") return String(value);
  if (typeof value === "string") return quote(value);
  if (Array.isArray(value)) return formatArray(value, indent, depth);
  if (isVonObject(value)) return formatObject(value, indent, depth);
  throw new VonError("unsupported VON value");
}

function formatArray(
  values: VonValue[],
  indent: number | undefined,
  depth: number,
): string {
  if (values.length === 0) return "[]";
  if (indent === undefined) {
    return `[${values.map((value) => formatValue(value, undefined, depth)).join(",")}]`;
  }
  const prefix = " ".repeat(indent * (depth + 1));
  const suffix = " ".repeat(indent * depth);
  const body = values
    .map((value) => `${prefix}${formatValue(value, indent, depth + 1)}`)
    .join(",\n");
  return `[\n${body}\n${suffix}]`;
}

function formatObject(
  entries: { [key: string]: VonValue },
  indent: number | undefined,
  depth: number,
): string {
  const keys = Object.keys(entries).sort();
  if (keys.length === 0) return "{}";
  if (indent === undefined) {
    return `{${keys
      .map((key) => `${formatKey(key)}:${formatValue(entries[key]!, undefined, depth)}`)
      .join(",")}}`;
  }
  const prefix = " ".repeat(indent * (depth + 1));
  const suffix = " ".repeat(indent * depth);
  const body = keys
    .map(
      (key) =>
        `${prefix}${formatKey(key)}: ${formatValue(entries[key]!, indent, depth + 1)}`,
    )
    .join(",\n");
  return `{\n${body}\n${suffix}}`;
}

function formatKey(key: string): string {
  if (isIdentifier(key) && key !== "null" && key !== "true" && key !== "false") {
    return key;
  }
  return quote(key);
}

function isIdentifier(key: string): boolean {
  if (key.length === 0) return false;
  if (!isIdentifierStart(key[0]!)) return false;
  for (let i = 1; i < key.length; i += 1) {
    if (!isIdentifierContinue(key[i]!)) return false;
  }
  return true;
}

function quote(value: string): string {
  let output = '"';
  for (const ch of value) {
    switch (ch) {
      case '"':
        output += '\\"';
        break;
      case "\\":
        output += "\\\\";
        break;
      case "\n":
        output += "\\n";
        break;
      case "\r":
        output += "\\r";
        break;
      case "\t":
        output += "\\t";
        break;
      default:
        output += ch;
    }
  }
  return `${output}"`;
}

File: projects/von.rs/von/src/format.rs
//! Canonical VON text formatting for Spark meta files.

use std::fmt::Write;

use serde::Serialize;

use crate::{VonError, VonValue, to_value};

/// Serializes a value to compact, deterministic VON text.
pub fn to_string<T>(value: &T) -> Result<String, VonError>
where
    T: Serialize,
{
    let value = to_value(value)?;
    Ok(format_value(&value, None, 0))
}

/// Serializes a value to deterministic, two-space-indented VON text.
pub fn to_string_indented<T>(value: &T) -> Result<String, VonError>
where
    T: Serialize,
{
    let value = to_value(value)?;
    Ok(format_value(&value, Some(2), 0))
}

fn format_value(value: &VonValue, indent: Option<usize>, depth: usize) -> String {
    match value {
        VonValue::Null => "null".to_owned(),
        VonValue::Bool(value) => value.to_string(),
        VonValue::Number(value) => value.to_string(),
        VonValue::String(value) => quote(value),
        VonValue::Array(values) => format_array(values, indent, depth),
        VonValue::Object(entries) => format_object(entries, indent, depth),
    }
}

fn format_array(values: &[VonValue], indent: Option<usize>, depth: usize) -> String {
    if values.is_empty() {
        return "[]".to_owned();
    }
    match indent {
        None => format!(
            "[{}]",
            values
                .iter()
                .map(|value| format_value(value, None, depth))
                .collect::<Vec<_>>()
                .join(",")
        ),
        Some(width) => {
            let prefix = " ".repeat(width * (depth + 1));
            let suffix = " ".repeat(width * depth);
            format!(
                "[\n{}\n{}]",
                values
                    .iter()
                    .map(|value| format!("{prefix}{}", format_value(value, indent, depth + 1)))
                    .collect::<Vec<_>>()
                    .join(",\n"),
                suffix
            )
        }
    }
}

fn format_object(
    entries: &std::collections::BTreeMap<String, VonValue>,
    indent: Option<usize>,
    depth: usize,
) -> String {
    if entries.is_empty() {
        return "{}".to_owned();
    }
    match indent {
        None => format!(
            "{{{}}}",
            entries
                .iter()
                .map(|(key, value)| format!(
                    "{}:{}",
                    format_key(key),
                    format_value(value, None, depth)
                ))
                .collect::<Vec<_>>()
                .join(",")
        ),
        Some(width) => {
            let prefix = " ".repeat(width * (depth + 1));
            let suffix = " ".repeat(width * depth);
            format!(
                "{{\n{}\n{}}}",
                entries
                    .iter()
                    .map(|(key, value)| format!(
                        "{prefix}{}: {}",
                        format_key(key),
                        format_value(value, indent, depth + 1)
                    ))
                    .collect::<Vec<_>>()
                    .join(",\n"),
                suffix
            )
        }
    }
}

fn format_key(key: &str) -> String {
    if is_identifier(key) && !matches!(key, "null" | "true" | "false") {
        key.to_owned()
    } else {
        quote(key)
    }
}

fn is_identifier(key: &str) -> bool {
    let mut chars = key.chars();
    chars.next().is_some_and(von_parser::is_identifier_start)
        && chars.all(von_parser::is_identifier_continue)
}

fn quote(value: &str) -> String {
    let mut output = String::with_capacity(value.len() + 2);
    output.push('"');
    for ch in value.chars() {
        match ch {
            '"' => output.push_str("\\\""),
            '\\' => output.push_str("\\\\"),
            '\n' => output.push_str("\\n"),
            '\r' => output.push_str("\\r"),
            '\t' => output.push_str("\\t"),
            _ => output
                .write_char(ch)
                .expect("writing to String cannot fail"),
        }
    }
    output.push('"');
    output
}
